src/operators/connect_neighborhood.c

```diff
@@ -162,3 +163,128 @@ igraph_error_t igraph_connect_neighborhood(igraph_t *graph, igraph_integer_t ord
 
     return IGRAPH_SUCCESS;
 }
+
+/**
+ * \function igraph_graph_power
+ * \brief Connect each vertex to its neighborhood.
+ *
+ * \experimental
+ *
+ * This function adds new edges to the input graph. Each vertex is connected
+ * to all vertices reachable by at most \p order steps from it
+ * (unless a connection already existed).  In other words, the \p order power of
+ * the graph is computed.
+ *
+ * </para><para> For undirected graphs reachability is always
+ * symmetric: if vertex A can be reached from vertex B in at
+ * most \p order steps, then the opposite is also true. Only one
+ * undirected (A,B) edge will be added in this case.
+ *
+ * \param graph The input graph.
+ * \param res The resulting graph.
+ * \param order Integer constant, it gives the distance within which
+ *    the vertices will be connected to the source vertex.
+ * \return Error code.
+ *
+ * Time complexity: O(|V|*d^k), |V| is the number of vertices in the
+ * graph, d is the average degree and k is the \p order argument.
+ */
+igraph_error_t igraph_graph_power(const igraph_t *graph, igraph_t *res,
+                                  igraph_integer_t order) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_dqueue_int_t q;
+    igraph_vector_int_t edges;
+    igraph_integer_t i, j, in;
+    igraph_integer_t *added;
+    igraph_vector_int_t neis;
+
+    if (order < 0) {
+        IGRAPH_ERRORF("Order cannot be negative, found %" IGRAPH_PRId ".",
+                IGRAPH_EINVAL, order);
+    }
+
+    if (order == 0) {
+        IGRAPH_CHECK(igraph_empty(res, igraph_vcount(graph), igraph_is_directed(graph)));
+        if (igraph_has_attribute_table()) {
+            IGRAPH_I_ATTRIBUTE_DESTROY(res);
+            IGRAPH_I_ATTRIBUTE_COPY(res, graph, 1, 1, 0);
+        }
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_copy(res, graph));
```

Comment:
I forgot to fix this part (only copy the necessary parts, instead of copying and removing), so I'm not actually finished yet.